_posts/2020-05-02-schedule-your-functions.md

```diff
@@ -0,0 +1,264 @@
+
+```yml
+version: 1.0
+provider:
+  name: openfaas
+  gateway: <gateway_url>
+functions:
+  lockbot:
+    lang: python3-flask-debian
+    handler: ./lockbot
+    image: <your_dockerhub_username>/lockbot:latest
+    environment:
+      github_username: <github_username>
+      github_repository: <repository_to_watch>
+      days_before_outdated: 90
+    secrets:
+      - auth-token
+    annotations:
+      topic: cron-function
+      schedule: "0 0 * * *"
+```
+
+### Augment the code
+
+Inside the `lockbot` folder you can see the `handler.py` file which contains the `handle` method which is the entry point to your function. Replace what's inside with the lockbot's code:
+
+```python
+import os
+import requests
+from datetime import datetime
+
+
+def get_difference_in_days(string_date):
+    # GitHub date format: 2019-01-13T21:17:09Z
+    created_date = string_date.split('T')[0].split('-')
+    for i in range(0, len(created_date)):
+        created_date[i] = int(created_date[i])
+    created_date = datetime(created_date[0], created_date[1], created_date[2])
+    return datetime.now() - created_date
+
+
+def lock_issue(issue_number, auth_header, username, repo):
+    r = requests.put(
+        url=f"https://api.github.com/repos/{username}/{repo}/issues/{issue_number}/lock", headers=auth_header)
+    if r.status_code == 204:
+        return f"Issue: {issue_number} locked"
+
+
+def get_authorization_header():
+    with open("/var/openfaas/secrets/auth-token") as file:
+        return {'Authorization': f"token {file.read().strip()}"}
+
+
+def handle(req):
+    username = os.getenv("github_username")
+    repo = os.getenv("github_repository")
+    days_before_outdated = int(os.getenv("days_before_outdated"))
+    r = requests.get(f"https://api.github.com/repos/{username}/{repo}/issues")
+    issues = r.json()
+
+    issue_numbers = []
+    for issue in issues:
+        if not issue.get('pull_request'):
+            diff = get_difference_in_days(issue.get('created_at'))
+            if diff.days > days_before_outdated and not issue.get("locked"):
+                issue_numbers.append(int(issue.get('number')))
+
+    locked_issues = ""
+    if len(issue_numbers) > 0:
+        auth_header = get_authorization_header()
+        for issue_number in issue_numbers:
+            locked_issues = locked_issues + \
+                f"{lock_issue(issue_number, auth_header, username, repo)}\n"
+
+    if locked_issues != "":
+        return locked_issues
+
+    return "No issues locked"
+```
+
+In general the function will fetch the issues from the repository you chose in the configuration section then it will filter them and calculate, using the date when they were opened, whether they are older than the 90 days we chose. If this is the case the lockbot will authenticate itself against the GitHub API with the access token we generated and will lock the issue from further conversation. The lockbot will respond with the locked issues or the lack of such.
```

Comment:
"In general" - I don't think this term is needed or adds anything.

Reply:
Will remove, was telling the reader to expect what the function does `in general`

Reply:
Removed :heavy_check_mark: